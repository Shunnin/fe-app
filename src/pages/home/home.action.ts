import { createAsyncActionType, createAsyncActionCreator } from '../../common/redux';

import { HOME_MODULE } from './home.constant';

export const GET_CURRENT_WEATHER = createAsyncActionType('GET_CURRENT_WEATHER', HOME_MODULE);
export const GET_DAILY_FORECAST = createAsyncActionType('GET_DAILY_FORECAST', HOME_MODULE);
export const GET_LOCATION = createAsyncActionType('GET_LOCATION', HOME_MODULE);

export const getCurrentWeather = createAsyncActionCreator(GET_CURRENT_WEATHER);
export const getDailyForecast = createAsyncActionCreator(GET_DAILY_FORECAST);
export const getLocation = createAsyncActionCreator(GET_LOCATION);
