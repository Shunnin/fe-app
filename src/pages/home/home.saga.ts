import axios from 'axios';
import { all, call, put, takeLatest } from 'redux-saga/effects';
import { reduce, join } from 'lodash-es';

import {
  INT_ZERO,
  DEFAULT_LOCALE,
  API_URL_GET_CURRENT_WEATHER,
  API_URL_GET_DAILY_FORECAST,
  API_URL_SEARCH_LOCATION,
  WEATHER_API_KEY,
} from '../../common/utility/constant';
import { UrlUtil } from '../../common/utility';

import {
  GET_CURRENT_WEATHER,
  GET_DAILY_FORECAST,
  GET_LOCATION,
  getCurrentWeather,
  getDailyForecast,
  getLocation,
} from './home.action';

const getCurrentWeatherApi = (value: any) => {
  const apiUrl = UrlUtil.buildQueryUrl(API_URL_GET_CURRENT_WEATHER, {
    ...value,
    appid: WEATHER_API_KEY,
  });

  return axios.get(apiUrl);
};

const getDailyForecastApi = (value: any) => {
  const apiUrl = UrlUtil.buildQueryUrl(API_URL_GET_DAILY_FORECAST, {
    ...value,
    appid: WEATHER_API_KEY,
  });

  return axios.get(apiUrl);
};

const searchLocationApi = (value: any) => {
  const data = JSON.stringify({
    query: value,
    type: 'city',
    language: DEFAULT_LOCALE,
  });

  return axios.post(API_URL_SEARCH_LOCATION, data);
};

export interface ResponseGenerator {
  config?: any;
  data?: any;
  headers?: any;
  request?: any;
  status?: number;
  statusText?: string;
}

const getCurrentWeatherSaga = function* (action) {
  try {
    const response: ResponseGenerator = yield call(getCurrentWeatherApi, action.payload);

    yield put(getCurrentWeather.success(response.data));
  } catch (e) {
    yield put(getCurrentWeather.error(e));
  }
};

const getDailyForecastSaga = function* (action) {
  try {
    const response: ResponseGenerator = yield call(getDailyForecastApi, action.payload);

    yield put(getDailyForecast.success(response.data));
  } catch (e) {
    yield put(getDailyForecast.error());
  }
};

const getLocationSaga = function* (action) {
  try {
    const response: ResponseGenerator = yield call(searchLocationApi, action.payload);
    const { data: { hits } = {} } = response;
    const locations = reduce(
      hits,
      (result, item) => {
        if (item.is_city) {
          result.push(join([item.locale_names[INT_ZERO], item.country], ', '));
        }

        return result;
      },
      [],
    );

    yield put(getLocation.success(locations));
  } catch (e) {
    yield put(getLocation.error());
  }
};

export const homeSaga = function* () {
  try {
    yield all([
      takeLatest(GET_CURRENT_WEATHER.REQUEST, getCurrentWeatherSaga),
      takeLatest(GET_DAILY_FORECAST.REQUEST, getDailyForecastSaga),
      takeLatest(GET_LOCATION.REQUEST, getLocationSaga),
    ]);
  } catch (e) {
    // ...
  }
};
