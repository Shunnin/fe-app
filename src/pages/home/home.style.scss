@import 'style/foundation/config';

.weather-section {
  margin-bottom: $b-space-2x;

	& &__forecast {
		margin-top: $b-space-2x;
		background-color: $c-wh;
		opacity: 0.75;
		box-shadow: 1px 1px 3px rgba(0, 0, 0, 0.1);
		border-radius: $b-border-radius-md;
		padding: $b-space-2x;
		overflow: hidden;

    &-title {
      border-bottom: 1px solid $c-gr3;
      padding-bottom: 1.5 * $b-space;
    }

		&-items {
			display: flex;
			justify-content: space-between;
			overflow-x: auto;
      margin-top: $b-space;

			&:last-child {
				margin-right: 0;
			}
		}

    @include respondto(xs, sm) {
			padding: $b-space;

      &-title {
        padding-bottom: $b-space-s;
      }

      &-items {
        margin-top: $b-space-s;
      }
		}
	}

  @include respondto(xs) {
    .weather-card {
      &__status {
        &-name {
          font-size: $b-font-size;
        }

        .temperature {
          font-size: 5 * $b-font-size;
        }
      }
    }

    & &__forecast {
      &-items {
        flex-direction: column;

        .weather-item {
          flex-direction: row;
          justify-content: space-around;
          margin-top: 0;
        }
      }
    }
  }
}
