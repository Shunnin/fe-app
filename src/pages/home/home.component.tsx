import { useEffect, FC, useState, memo, useCallback } from 'react';
import { connect } from 'react-redux';
import { isEmpty, map } from 'lodash-es';
import { FormattedMessage, useIntl } from 'react-intl';

import { DailyForecastData } from '../../models';
import { WeatherCard, WeatherItem, composeWeatherData, composeDailyForecastData } from '../../common/components/shared';
import { Header, SearchBox, Loading } from '../../common/components/core';
import { changeTempUnit as changeTempUnitAction } from '../../common/redux/service/app/';
import { pipe } from '../../common/utility';

import { getCurrentWeather, getDailyForecast, getLocation } from './home.action';
import { homeSelectors } from './home.selector';
import { homeReducer } from './home.reducer';
import { homeSaga } from './home.saga';
import './home.style';

interface IDailyForecastProps {
  dailyForeCast: DailyForecastData[];
}

const DailyForeCast: FC<IDailyForecastProps> = memo(({ dailyForeCast }) => {
  if (isEmpty(dailyForeCast)) {
    return <></>;
  }

  return (
    <div className="weather-section__forecast">
      <div className="weather-section__forecast-title text-uppercase">
        <FormattedMessage defaultMessage="7-day forecast" />
      </div>
      <div className="weather-section__forecast-items">
        {map(dailyForeCast, (item: DailyForecastData, index: number) => {
          const {
            day,
            id,
            forecast,
            tempDetails: { tempMax, tempMin },
          } = item;

          return (
            <WeatherItem
              key={index}
              day={day}
              weatherCode={id}
              tempMax={tempMax}
              tempMin={tempMin}
              forecast={forecast}
            />
          );
        })}
      </div>
    </div>
  );
});
DailyForeCast.displayName = 'DailyForeCast';

interface IHomeProps {
  error: any;
  loading: boolean;
  weather: any;
  dailyForecast: any;
  locations: string[];
  degreeType: string;
  getCurrentWeatherRequest: Function;
  getDailyForecastRequest: Function;
  getLocationRequest: Function;
  changeTempUnit: Function;
}

const Home: FC<IHomeProps> = ({
  // Props
  error,
  loading,
  weather,
  dailyForecast,
  locations,
  degreeType,
  // Actions
  getCurrentWeatherRequest,
  getDailyForecastRequest,
  getLocationRequest,
  changeTempUnit,
}) => {
  const intl = useIntl();
  const [queryParams, setQueryParams] = useState(null);
  const [composedWeather, setComposedWeather] = useState({});
  const [composedDailyForeCast, setComposedDailyForeCast] = useState([]);

  const handleChangeTempUnit = useCallback(value => changeTempUnit(value), [changeTempUnit]);

  const handleClickSearch = useCallback(value => getLocationRequest(value), [getLocationRequest]);

  const handleClickSuggestion = useCallback(value => setQueryParams({ q: value }), [setQueryParams]);

  useEffect(() => {
    if (!isEmpty(queryParams)) {
      getCurrentWeatherRequest(queryParams);
      getDailyForecastRequest(queryParams);
    }
  }, [queryParams]);

  useEffect(() => {
    if (isEmpty(weather)) {
      setComposedWeather([]);
    } else {
      const result = composeWeatherData(weather);

      setComposedWeather(result);
    }
  }, [weather]);

  useEffect(() => {
    if (isEmpty(dailyForecast)) {
      setComposedDailyForeCast([]);
    } else {
      const result = composeDailyForecastData(dailyForecast);

      setComposedDailyForeCast(result);
    }
  }, [dailyForecast]);

  return (
    <>
      {loading && <Loading />}
      <div className="container">
        <Header />
        <div className="weather-section">
          <SearchBox
            placeholder={intl.formatMessage({ defaultMessage: 'Search for location' })}
            suggestions={locations}
            onClickSearch={handleClickSearch}
            onClickSuggestion={handleClickSuggestion}
          />
          {isEmpty(error) ? (
            <>
              <WeatherCard
                weather={composedWeather}
                degreeType={degreeType}
                onClickChangeTempUnit={handleChangeTempUnit}
              />
              <DailyForeCast dailyForeCast={composedDailyForeCast} />
            </>
          ) : (
            <div className="text-center">
              <FormattedMessage defaultMessage="Sorry! No results found" />
            </div>
          )}
        </div>
      </div>
    </>
  );
};

export default pipe(
  connect(homeSelectors, {
    getCurrentWeatherRequest: getCurrentWeather.request,
    getDailyForecastRequest: getDailyForecast.request,
    getLocationRequest: getLocation.request,
    changeTempUnit: changeTempUnitAction,
  }),
)(Home);

export const dependencies = {
  reducer: homeReducer,
  saga: homeSaga,
};
