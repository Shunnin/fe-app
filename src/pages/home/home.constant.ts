import { IDefaultAction } from '../../common/redux';

import { GET_CURRENT_WEATHER, GET_DAILY_FORECAST, GET_LOCATION } from './home.action';

export const HOME_MODULE = 'homeModule';

export interface IWeatherState {
  loading: boolean;
  weather: any;
  dailyForecast: any;
  locations: any[];
  error: any;
}

// #region IPayload
export interface ISearchWeatherPayload {
  q?: string;
}

// #region IPayload
export interface ISearchWeatherErrorPayload {
  error: any;
}

export interface IGetCurrentWeatherSuccessPayload {
  weather: any;
}

export interface IGetDailyForecastSuccessPayload {
  dailyForecast: any;
}

export interface IGetLocationSuccessPayload {
  locations: string[];
}

export interface IGetLocationRequestPayload {
  keyword: string;
}
// #endregion IPayload

// #region IAction
export interface IGetCurrentWeatherRequest extends IDefaultAction {
  type: typeof GET_CURRENT_WEATHER.REQUEST;
  payload: ISearchWeatherPayload;
}

export interface IGetCurrentWeatherSuccess extends IDefaultAction {
  type: typeof GET_CURRENT_WEATHER.SUCCESS;
  response: IGetCurrentWeatherSuccessPayload;
}

export interface IGetCurrentWeatherError extends IDefaultAction {
  type: typeof GET_CURRENT_WEATHER.ERROR;
  error: ISearchWeatherErrorPayload;
}

export interface IGetDailyForecastRequest extends IDefaultAction {
  type: typeof GET_DAILY_FORECAST.REQUEST;
  payload: ISearchWeatherPayload;
}

export interface IGetDailyForecastSuccess extends IDefaultAction {
  type: typeof GET_DAILY_FORECAST.SUCCESS;
  response: IGetDailyForecastSuccessPayload;
}

export interface IGetDailyForecastError extends IDefaultAction {
  type: typeof GET_DAILY_FORECAST.ERROR;
}

export interface IGetLocationRequest extends IDefaultAction {
  type: typeof GET_LOCATION.REQUEST;
  payload: IGetLocationRequestPayload;
}

export interface IGetLocationSuccess extends IDefaultAction {
  type: typeof GET_LOCATION.SUCCESS;
  response: IGetLocationSuccessPayload;
}

export interface IGetLocationError extends IDefaultAction {
  type: typeof GET_LOCATION.ERROR;
}
// #endregion IAction

export type WeatherActions =
  | IGetCurrentWeatherRequest
  | IGetCurrentWeatherSuccess
  | IGetCurrentWeatherError
  | IGetDailyForecastRequest
  | IGetDailyForecastSuccess
  | IGetDailyForecastError
  | IGetLocationRequest
  | IGetLocationSuccess
  | IGetLocationError;
