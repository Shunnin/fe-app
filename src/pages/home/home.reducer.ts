import { WeatherActions, IWeatherState } from './home.constant';
import { GET_CURRENT_WEATHER, GET_DAILY_FORECAST, GET_LOCATION } from './home.action';

const initialState: IWeatherState = {
  loading: false,
  weather: {},
  dailyForecast: {},
  locations: [],
  error: null,
};

export const homeReducer = (state = initialState, action: WeatherActions) => {
  const { type, response, error } = action;

  switch (type) {
    case GET_CURRENT_WEATHER.REQUEST:
      return {
        ...state,
        loading: true,
      };

    case GET_CURRENT_WEATHER.SUCCESS:
      return {
        ...state,
        loading: false,
        weather: response,
        error: null,
      };

    case GET_CURRENT_WEATHER.ERROR:
      return {
        ...state,
        loading: false,
        weather: {},
        dailyForecast: {},
        error: error,
      };

    case GET_DAILY_FORECAST.REQUEST:
      return {
        ...state,
        loading: true,
      };

    case GET_DAILY_FORECAST.SUCCESS:
      return {
        ...state,
        loading: false,
        dailyForecast: response,
      };

    case GET_DAILY_FORECAST.ERROR:
      return {
        ...state,
        loading: false,
        dailyForecast: [],
      };

    case GET_LOCATION.SUCCESS:
      return {
        ...state,
        locations: response,
      };

    case GET_LOCATION.ERROR:
      return {
        ...state,
        locations: [],
      };

    default:
      return {
        ...state,
      };
  }
};
