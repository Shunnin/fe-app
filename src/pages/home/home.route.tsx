import { store } from '../../store';
import { asyncComponent } from '../../common/components/core';
import { URL_HOME } from '../../common/utility/constant';

import { HOME_MODULE } from './home.reducer';

const component = asyncComponent({
  component: () =>
    import(
      /* webpackChunkName: 'front-app' */
      /* webpackMode: "lazy" */
      /* webpackPrefetch: true */
      './home.component'
    ),
  store,
  moduleName: HOME_MODULE,
});

export const homeRoute = {
  component,
  path: URL_HOME,
  exact: true,
  routes: [],
};
