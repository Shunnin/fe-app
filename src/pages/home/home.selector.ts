import { createSelector } from 'reselect';

import { HOME_MODULE } from './home.reducer';

interface IState {
  homeModule?: {
    loading: boolean;
    weather: any;
    dailyForecast: any;
    locations: string[];
    errors: string;
  };
}

const getLoading = (state: IState) => state[HOME_MODULE].loading;

const getWeather = (state: IState) => state[HOME_MODULE].weather;

const getDailyForecast = (state: IState) => state[HOME_MODULE].dailyForecast;

const getLocations = (state: IState) => state[HOME_MODULE].locations;

const getErrors = (state: IState) => state[HOME_MODULE].errors;

export const getWeatherSelector = createSelector(getWeather, weather => weather);

export const getDailyForecastSelector = createSelector(getDailyForecast, dailyForecast => dailyForecast);

export const getLocationsSelector = createSelector(getLocations, (locations: string[]) => locations);

export const getLoadingSelector = createSelector(getLoading, loading => loading);

export const getErrorSelector = createSelector(getErrors, errors => errors);

// import { createSelector, combineSelectors } from '../../common/redux';

// const homeStateSelector = state => state.get(HOME_MODULE);

// export const weatherSelector = createSelector(homeStateSelector, moduleState => {
//   return moduleState.get('weather');
// });

// export const dailyForecastSelector = createSelector(homeStateSelector, moduleState => {
//   return moduleState.get('dailyForecast');
// });

// export const locationsSelector = createSelector(homeStateSelector, moduleState => {
//   return moduleState.get('locations');
// });

// export const loadingSelector = createSelector(homeStateSelector, moduleState => {
//   return moduleState.get('loading');
// });

// export const errorsSelector = createSelector(homeStateSelector, moduleState => {
//   return moduleState.get('errors');
// });

// export const homeSelector = combineSelectors({
//   weather: getWeatherSelector,
//   dailyForecast: getDailyForecastSelector,
//   locations: getLocationsSelector,
//   loading: getLoadingSelector,
//   errors: getErrorSelector,
// });
