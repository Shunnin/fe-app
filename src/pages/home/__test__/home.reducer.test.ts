import { homeReducer } from '../home.reducer';
import { getCurrentWeather, getDailyForecast, getLocation } from '../home.action';

describe('Home Reducer', () => {
  it('returns the initial state', () => {
    const action = {
      type: 'DUMMY_ACTION',
    };
    const reducer = homeReducer(undefined, action);

    expect(reducer).toMatchSnapshot();
  });

  describe('handles the getCurrentWeather action', () => {
    it('returns loading state is true with execute action request', () => {
      const getCurrentWeatherAction = getCurrentWeather.request({ q: 'ho chi minh' });
      const reducer = homeReducer(undefined, getCurrentWeatherAction);

      expect(reducer).toMatchSnapshot();
    });

    it('returns weather state and loading is false with execute action success', () => {
      const getCurrentWeatherAction = getCurrentWeather.success({ place: 'ho chi minh' });
      const reducer = homeReducer(undefined, getCurrentWeatherAction);

      expect(reducer).toMatchSnapshot();
    });

    it('returns empty weather and loading is false is false with execute action error', () => {
      const getCurrentWeatherAction = getCurrentWeather.error('Can not fetch the data');
      const reducer = homeReducer(undefined, getCurrentWeatherAction);

      expect(reducer).toMatchSnapshot();
    });
  });

  describe('handles the getDailyForecast action', () => {
    it('returns loading state is true with execute action request', () => {
      const getDailyForecastAction = getDailyForecast.request({ q: 'ho chi minh' });
      const reducer = homeReducer(undefined, getDailyForecastAction);

      expect(reducer).toMatchSnapshot();
    });

    it('returns dailyForecast state and loading is false with execute action success', () => {
      const getDailyForecastAction = getDailyForecast.success({ place: 'ho chi minh', temp: 30 });
      const reducer = homeReducer(undefined, getDailyForecastAction);

      expect(reducer).toMatchSnapshot();
    });

    it('returns empty dailyForecast and loading is false is false with execute action error', () => {
      const getDailyForecastAction = getDailyForecast.error('Can not fetch the data');
      const reducer = homeReducer(undefined, getDailyForecastAction);

      expect(reducer).toMatchSnapshot();
    });
  });

  describe('handles the getLocation action', () => {
    it('returns the initial state with execute action request', () => {
      const getLocationAction = getLocation.request({ q: 'ho chi minh' });
      const reducer = homeReducer(undefined, getLocationAction);

      expect(reducer).toMatchSnapshot();
    });

    it('returns locations state with execute action success', () => {
      const getLocationAction = getLocation.success(['ho chi minh']);
      const reducer = homeReducer(undefined, getLocationAction);

      expect(reducer).toMatchSnapshot();
    });

    it('returns empty locations with execute action error', () => {
      const getLocationAction = getLocation.error('Can not fetch the data');
      const reducer = homeReducer(undefined, getLocationAction);

      expect(reducer).toMatchSnapshot();
    });
  });
});
