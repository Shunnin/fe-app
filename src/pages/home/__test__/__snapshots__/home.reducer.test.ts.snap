// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Reducer handles the getCurrentWeather action returns empty weather and loading is false is false with execute action error 1`] = `
Object {
  "dailyForecast": Object {},
  "error": "Can not fetch the data",
  "loading": false,
  "locations": Array [],
  "weather": Object {},
}
`;

exports[`Home Reducer handles the getCurrentWeather action returns loading state is true with execute action request 1`] = `
Object {
  "dailyForecast": Object {},
  "error": null,
  "loading": true,
  "locations": Array [],
  "weather": Object {},
}
`;

exports[`Home Reducer handles the getCurrentWeather action returns weather state and loading is false with execute action success 1`] = `
Object {
  "dailyForecast": Object {},
  "error": null,
  "loading": false,
  "locations": Array [],
  "weather": Object {
    "place": "ho chi minh",
  },
}
`;

exports[`Home Reducer handles the getDailyForecast action returns dailyForecast state and loading is false with execute action success 1`] = `
Object {
  "dailyForecast": Object {
    "place": "ho chi minh",
    "temp": 30,
  },
  "error": null,
  "loading": false,
  "locations": Array [],
  "weather": Object {},
}
`;

exports[`Home Reducer handles the getDailyForecast action returns empty dailyForecast and loading is false is false with execute action error 1`] = `
Object {
  "dailyForecast": Array [],
  "error": null,
  "loading": false,
  "locations": Array [],
  "weather": Object {},
}
`;

exports[`Home Reducer handles the getDailyForecast action returns loading state is true with execute action request 1`] = `
Object {
  "dailyForecast": Object {},
  "error": null,
  "loading": true,
  "locations": Array [],
  "weather": Object {},
}
`;

exports[`Home Reducer handles the getLocation action returns empty locations with execute action error 1`] = `
Object {
  "dailyForecast": Object {},
  "error": null,
  "loading": false,
  "locations": Array [],
  "weather": Object {},
}
`;

exports[`Home Reducer handles the getLocation action returns locations state with execute action success 1`] = `
Object {
  "dailyForecast": Object {},
  "error": null,
  "loading": false,
  "locations": Array [
    "ho chi minh",
  ],
  "weather": Object {},
}
`;

exports[`Home Reducer handles the getLocation action returns the initial state with execute action request 1`] = `
Object {
  "dailyForecast": Object {},
  "error": null,
  "loading": false,
  "locations": Array [],
  "weather": Object {},
}
`;

exports[`Home Reducer returns the initial state 1`] = `
Object {
  "dailyForecast": Object {},
  "error": null,
  "loading": false,
  "locations": Array [],
  "weather": Object {},
}
`;
